File: original/parsechart.js
/* Script for reading the bms file */

let chartinfo;
const fr = new FileReader();
let bars = [];
const maxlen = 10000;
// example of Chart information
// #GENRE MUSIC
// #TITLE Agito
// #ARTIST polycube ~ øc(alpha complex)
// #BPM 172
// #PLAYLEVEL 3
// #RANK 3
class Chart{
    constructor(){
        this.genre = "";
        this.title = "";
        this.artist = "";
        this.bpm = 0;
        this.playlevel = 0;
        this.rank = 0;
        this.maxbeat = 0;
        this.notecount = 0;
    }
}
class Bar{
    constructor(){
        this.notes = []; // an order list of the groups of notes
        // with the respective channels to play them and time to play them
        this.sigchange = false;
        this.sigvalue = 1; // 4/4
        this.bpmchange = false;
        this.bpmvalue = 0;
        this.stop = false;
        this.stopvalue = 0;
    }
}
class Hold{
    constructor(){
        this.channel = 0;
        this.start = 0;
        this.length = 0;
    }
}
function readchart() {
    fetch(charturl).
        then(response => {
            if (response.ok){
                return response.text();
            }
            else {
                throw new Error("Map not found error! status: "+ response.status);
            }
        }).
        then(text => {
            let chartdata = text;
            chartinfo = parsechart(chartdata);
        });
        return;
}

function parsechart(chartdata){
    let chartinfo, splitted, header, expansion, main, maxbeat;
    // separate the array 
    // when found *----------------------
    splitted = chartdata.split("*----------------------");
    header = splitted[1];
    if (chartdata.match(/EXPANSION FIELD/)){
        expansion = splitted[2];
        main = splitted[3];
    }
    else{
        main = splitted[2];
    }

    // get the information of the Chart
    chartinfo = new Chart();
    chartinfo.genre = header.match(/\#GENRE .+/)[0].slice(7);
    chartinfo.artist = header.match(/\#ARTIST .+/)[0].slice(8);
    chartinfo.title = header.match(/\#TITLE .+/)[0].slice(7);
    chartinfo.bpm = Number(header.match(/\#BPM \d+/)[0].slice(5));
    chartinfo.playlevel = Number(header.match(/\#PLAYLEVEL \d+/)[0].slice(11));
    chartinfo.rank = Number(header.match(/\#RANK \d+/)[0].slice(6));
    try{
        chartinfo.subtitle = header.match(/\#SUBTITLE .+/)[0].slice(10);
    }
    catch{
        chartinfo.subtitle = "";
    }
    let bpms = {};
    // console.log(bpms)
    // match the #BPMxx N
    matchbpm = header.match(/\#BPM\d+ \d+(\.\d+)?/g);
    if (matchbpm != null){
        for (let i of matchbpm){
            bpms[i.slice(4,6)] = Number(i.slice(7));
        }
    }
    let stops = {};
    // match the #STOPxx N
    matchstop = header.match(/\#STOP\d+ \d+/g);
    if (matchstop != null){
        for (let i of matchstop){
            stops[i.slice(5,7)] = Number(i.slice(8));
        }
    }
    // audios = {};
    // // get the audio
    // for (let i of header.split("\r\n")){
    //     if (i.startsWith("#WAV")){
    //         audios[i.slice(4,6)] = new Audio(dir + i.slice(7, -3) + "ogg");
    //     }
    // }
    // get the number of beats
    for (let i of main.split("\r\n").reverse()){
        if (i.startsWith("#")){
            maxbeat = Number(i.slice(1,4));
            chartinfo.maxbeat = maxbeat;
            break;
        }
    }
    bars = [];
    // in every bar[i] there is an an order list of the groups of notes
    // with the respective channels to play them and time to play the next group

    let barelements = []; // auxiliar array to store the elements of the bar
    let bpm = chartinfo.bpm;
    let holdstart = [null, null, null, null, null, null, null];
    for (let i = 0; i <= maxbeat; i++){
        bars[i] = new Bar();
        barelements[i] = [];
    }
    for (let i of main.split("\r\n")){
        if (!i.startsWith("#")){
            continue;
        }
        barnumber = Number(i.slice(1,4));
        channel = i.slice(4,6);
        // if beat 02 beat is not 4/4
        // put in front the change of time signature
        if (i.slice(4,6) == "02"){
            bars[barnumber].sigchange = true;
            bars[barnumber].sigvalue = Number(i.slice(7));
            continue;
        }
        // bpm change
        if (i.slice(4,6) == "03"){
            bars[barnumber].bpmchange = true;
            objects = i.slice(7).match(/.{1,2}/g).map(x => parseInt(x, 16));
            // the index of first non null element
            let index = objects.findIndex(x => x != 0);
            bars[barnumber].bpmvalue = Math.max(...objects);
            bars[barnumber].sigvalue *= ((objects.length - index) + index * bars[barnumber].bpmvalue / bpm) / objects.length;
            bpm = bars[barnumber].bpmvalue;
            continue;
        }
        // bpm change 2
        if (i.slice(4,6) == "08"){
            bars[barnumber].bpmchange = true;
            bars[barnumber].bpmvalue = bpms[i.slice(7, 9)];
            bpm = bars[barnumber].bpmvalue;
        }
        // stop
        if (i.slice(4,6) == "09"){
            bars[barnumber].stop = true;
            bars[barnumber].stopvalue = stops[i.slice(7, 9)];
        }
        bars[barnumber].bpmvalue = bpm;
        // if it's click notes
        if (channel[0] == "1" || channel[0] == "5"){
            objects = i.slice(7).match(/.{1,2}/g);
            for (let j = 0; j < objects.length; j++){
                // if it's null
                if (objects[j] == "00"){
                    continue;
                }
                if (channel[1] > "7"){
                    channel = channel[0] + (channel[1] - 2);
                    // console.log(channel)
                }
                let hold = null;
                // if it's hold notes
                if (channel[0] == "5"){
                    // if it's the start of the hold note
                    if (holdstart[channel[1]] == null){
                        holdstart[channel[1]] = Number(barnumber) + j / objects.length;
                    }
                    // if it's the end of the hold note
                    else{
                        hold = new Hold();
                        hold.channel = channel;
                        hold.start = holdstart[channel[1]] % 1;
                        hold.length = 0;
                        // calculate the length of the hold note
                        // console.log(bars[Math.floor(holdstart[channel[1]])], holdstart[channel[1]]);
                        // console.log(holdstart[channel[1]]);
                        let tmpbpm = bars[Math.floor(holdstart[channel[1]])].bpmvalue;
                        // console.log(barnumber, bpm, tmpbpm, hold.length, bars[barnumber].bpmvalue, objects.length, holdstart[channel[1]]);
                        startlen = bars[Math.floor(holdstart[channel[1]])].sigvalue * hold.start * chartinfo.bpm / bars[[Math.floor(holdstart[channel[1]])]].bpmvalue;
                        for (let bar of bars.slice(holdstart[channel[1]], barnumber)){
                            if (bar.bpmchange == true){
                                tmpbpm = bar.bpmvalue;
                            }
                            hold.length += bar.sigvalue * chartinfo.bpm / tmpbpm;
                        }
                        hold.length += j / objects.length * bars[barnumber].sigvalue * chartinfo.bpm / bars[barnumber].bpmvalue - startlen;
                        // console.log(barnumber, bpm, tmpbpm, hold.length, bars[barnumber].bpmvalue, objects.length)
                        if (hold.length > maxlen){
                            hold.length = maxlen;
                        }
                        // console.log(hold, holdstart[channel[1]])
                        barelements[holdstart[channel[1]] - hold.start].push({
                            channel: channel,
                            object: objects[j],
                            hold: hold,
                            time: hold.start
                        });
                        holdstart[channel[1]] = null;     
                    }
                    continue;
                }
                barelements[barnumber].push({
                    channel: channel,
                    object: objects[j],
                    hold: hold,
                    time: j / objects.length
                });
                
            }
        }
        
    }
    // sort the elements of the bar by time
    for (let i of barelements){
        i.sort((a, b) => a.time - b.time);
    }
    // group the elements of the bar by time
    for (let i = 0; i < barelements.length; i++){
        let j = 0;
        while (j < barelements[i].length){
            let k = j;
            let time = barelements[i][j].time;
            let group = {
                time : time,
                channels: new Set(),
                holds: new Set()
            };
            while (k < barelements[i].length && barelements[i][k].time == time){
                if (barelements[i][k].channel[0] == "1"){
                    group.channels.add(barelements[i][k].channel);
                }
                else{
                    // console.log(barelements[i][k]);
                    group.holds.add(barelements[i][k].hold);
                }
                k++;
            }
            bars[i].notes.push(group);
            j = k;
        }
    }
    // count the total number of notes
    for (let i of bars){
        for (let j of i.notes){
            chartinfo.notecount += j.channels.size;
            chartinfo.notecount += j.holds.size * 2;
        }
    }

    return chartinfo;
}

File: src/parsechart.rs
// use std::io;
use std::collections::{HashMap, HashSet};
use std::str::Lines;
use std::fs;

// use std::error::Error;
mod classes;
use self::classes::{Chart, Bar, Note, NoteGroup, Hold};

// Final formatting of the output file
pub fn generate_output_file(chart: &Chart, bars: &Vec<Bar>, output_file: &str) {
    let mut output = String::new();
    output.push_str(&format!("{{\n\t\"chartinfo\": " ));
    output.push_str(&chart.to_string());
    output.push_str(&format!(",\n"));
    output.push_str(&format!("\n\"bars\": [\n"));
    for (i, bar) in bars.iter().enumerate() {
        output.push_str(&format!("\t{}", bar.to_string()));
        if i < bars.len() - 1 {
            output.push_str(",\n");
        } else {
            output.push_str("\n");
        }
    }
    output.push_str(&format!("]\n}}"));
    fs::write(output_file, output).expect("Unable to write file");

    println!("Output file: {}", output_file);
}

//  Sorts and groups notes by time
pub fn sort_and_group_notes(bars: &mut Vec<Bar>, barelements: &mut Vec<Vec<Note>>) {
    for i in 0..barelements.len() {
        barelements[i as usize].sort_by(|a, b| a.time.partial_cmp(&b.time).unwrap());
    }

    for i in 0..barelements.len() {
        let mut j = 0;

        while j < barelements[i as usize].len() {
            let mut k = j;
            let time = barelements[i as usize][j as usize].time;
            let mut group = NoteGroup {
                time: barelements[i as usize][j as usize].time,
                channels: HashSet::<u32>::new(),
                holds: HashSet::<Option<Hold>>::new(),
            };

            while k < barelements[i as usize].len() && barelements[i as usize][k as usize].time == time {
                if barelements[i as usize][k as usize].channel <= 20 {
                    group.channels.insert(barelements[i as usize][k as usize].channel);
                } else {
                    group.holds.insert(barelements[i as usize][k as usize].hold);
                }
                k += 1;
            }

            j = k;
            bars[i as usize].notes.push(group);
        }
    }
}

// ##########################################
// # Functions that process the chart data
// ##########################################

pub fn process_chart_info(lines: &mut std::str::Lines) -> (HashMap<String, f64>, HashMap<String, u32>, Chart) {
    let mut bpms: HashMap<String, f64> = HashMap::new();
    let mut stops: HashMap<String, u32> = HashMap::new();
    let mut chart = Chart {
        genre: String::new(),
        title: String::new(),
        artist: String::new(),
        bpm: 0.0,
        playlevel: 0,
        rank: 0,
        subtitle: String::new(),
    };

    let mut line = lines.next();
    let mut line_value = String::new();

    while line.is_none() == false && line_value.starts_with("*---------------------- MAIN DATA FIELD") == false {
        line_value = line.unwrap().to_string();

        let chart_info = line_value[..].split(" ").collect::<Vec<&str>>();

        let info = chart_info[0].to_string();
        let info_value = chart_info[1..].join(" ").to_string();
        if chart_info[0] == "#GENRE" {
            chart.genre = info_value.to_string();
        } else if chart_info[0] == "#TITLE" {
            chart.title = info_value.to_string();
        } else if chart_info[0] == "#ARTIST" {
            chart.artist = info_value.to_string();
        } else if chart_info[0] == "#BPM" {
            chart.bpm = chart_info[1].parse().unwrap();
        } else if chart_info[0] == "#PLAYLEVEL" {
            chart.playlevel = chart_info[1].parse().unwrap();
        } else if chart_info[0] == "#RANK" {
            chart.rank = chart_info[1].parse().unwrap();
        } else if chart_info[0] == "#SUBTITLE" {
            chart.subtitle = info_value.to_string();
        } else if chart_info[0].starts_with("#BPM") {
            bpms.insert(info[4..].to_string(), info_value.parse().unwrap());
        } else if chart_info[0].starts_with("#STOP") {
            stops.insert(info[5..].to_string(), info_value.parse().unwrap());
        }
        line = lines.next();
    }

    (bpms, stops, chart)
}

const MAXLEN: f64 = 1000.0;

pub fn process_main_data(main_start: &mut Lines, chart: &Chart, bpms: &HashMap<String, f64>, stops: &HashMap<String, u32>) -> (Vec<Bar>, Vec<Vec<Note>>) {
    let mut bars = Vec::<Bar>::new();
    let mut barelements = Vec::<Vec<Note>>::new();
    let bpm = chart.bpm;
    let mut holdstart = [0.0; 8];

    let mut nbars = 0;
    let mut lines = main_start.clone();

    let mut line = lines.next();
    let mut line_value;

    while line.is_none() == false {
        line_value = line.unwrap().to_string();
        if line_value.starts_with("#") == true {
            nbars = line_value[1..4].parse::<u32>().unwrap();
        }
        line = lines.next();
    }
    nbars += 1;

    lines = main_start.clone();
    line = lines.next();

    for _ in 0..nbars {
        bars.push(Bar::default());
        barelements.push(Vec::<Note>::new());
    }

    while line.is_none() == false {
        line_value = line.unwrap().to_string();

        if !line_value.starts_with("#") {
            line = lines.next();
            continue;
        }

        let barnumber = line_value[1..4].parse::<u32>().unwrap();
        let channel = line_value[4..6].parse::<u32>().unwrap();
        let value = line_value[7..].to_string();


        bars[barnumber as usize].bpmvalue = bpm;

        if channel == 2 {
            process_sigchange(&mut bars, barnumber, value);
        } else if channel == 3 {
            process_bpmchange(&mut bars, barnumber, value, &chart);
        } else if channel == 8 {
            process_bpmchange2(&mut bars, barnumber, value, &bpms);
        } else if channel == 9 {
            process_stop(&mut bars, barnumber, value, &stops);
        } else
        if (channel >= 11 && channel <= 19) || (channel >= 51 && channel <= 59) {
            process_note(&mut barelements, &mut holdstart, barnumber, channel, value, &bars, &chart);
        }

        line = lines.next();
    }

    for i in 0..barelements.len() {
        barelements[i as usize].sort_by(|a, b| a.time.partial_cmp(&b.time).unwrap());
    }

    (bars, barelements)
}


pub fn process_note(
    barelements: &mut Vec<Vec<Note>>,
    holdstart: &mut [f64; 8],
    barnumber: u32,
    mut channel: u32,
    value: String,
    bars: &Vec<Bar>,
    chart: &Chart,
) {
    // channel -= 1;

    match channel % 10 {
        7 | 8 | 9 => {
            channel -= 2;
        }
        _ => {}
    }

    let objects = value.chars().collect::<Vec<char>>();
    let objects = objects.chunks(2);
    let objects = objects.map(|x| x.iter().collect::<String>()).collect::<Vec<String>>();

    for j in 0..objects.len() {
        let object = objects[j].to_string();
        let mut hold = <Option<Hold>>::None;
        let mut starttime: f64 = j as f64 / objects.len() as f64;

        if object == "00" {
            continue;
        }

        // check if it's a hold
        if channel / 10 == 5 {
            // check if it's the start of the hold
            if holdstart[(channel % 10) as usize] == 0.0 {
                holdstart[(channel % 10) as usize] = barnumber as f64 + starttime;
            } else {
                let mut hold_unwrap = Hold {
                    channel: channel,
                    start: holdstart[(channel % 10) as usize] % 1.0,
                    length: 0.0,
                };
                
                let mut tmpbpm = bars[holdstart[(channel % 10) as usize].floor() as usize].bpmvalue;
                let startlen = bars[holdstart[(channel % 10) as usize].floor() as usize].sigvalue * hold_unwrap.start * chart.bpm / bars[holdstart[(channel % 10) as usize].floor() as usize].bpmvalue;

                for k in holdstart[(channel % 10) as usize].floor() as usize .. barnumber as usize {
                    if bars[k].bpmchange == true {
                        tmpbpm = bars[k].bpmvalue;
                    }
                    hold_unwrap.length += bars[k].sigvalue * chart.bpm / tmpbpm;
                }
                hold_unwrap.length += j as f64 / objects.len() as f64 * bars[barnumber as usize].sigvalue * chart.bpm / bars[barnumber as usize].bpmvalue - startlen;
                
                if hold_unwrap.length > MAXLEN {
                    hold_unwrap.length = MAXLEN;
                }

                holdstart[(channel % 10) as usize] = 0.0;

                starttime = hold_unwrap.start;
                hold = Some(hold_unwrap);
            }
        }

        barelements[barnumber as usize].push(Note {
            channel: channel,
            object: object,
            hold: hold,
            time: starttime as f64,
        });
    }
}

pub fn process_sigchange(bars: &mut Vec<Bar>, barnumber: u32, value: String) {
    bars[barnumber as usize].sigchange = true;
    bars[barnumber as usize].sigvalue = value.parse::<f64>().unwrap();
}

pub fn process_bpmchange(bars: &mut Vec<Bar>, barnumber: u32, value: String, chart: &Chart) {
    bars[barnumber as usize].bpmchange = true;
    let bpmvalues = value.chars().collect::<Vec<char>>();
    let bpmvalues = bpmvalues.chunks(2);
    let mut index: u32 = 0;
    let size = bpmvalues.len();

    for bpmvalue in bpmvalues {
        let bpmvalue = bpmvalue.iter().collect::<String>();
        if bpmvalue != "00" {
            bars[barnumber as usize].bpmvalue = u32::from_str_radix(&bpmvalue, 16).unwrap() as f64;
            break;
        }
        index += 1;
    }

    bars[barnumber as usize].sigvalue *= ((size as u32 - index) as f64 + index as f64 * bars[barnumber as usize].bpmvalue as f64 / chart.bpm as f64) / size as f64;
}

pub fn process_bpmchange2(bars: &mut Vec<Bar>, barnumber: u32, value: String, bpms: &HashMap<String, f64>) {
    bars[barnumber as usize].bpmchange = true;
    bars[barnumber as usize].bpmvalue = bpms[&value];
}

pub fn process_stop(bars: &mut Vec<Bar>, barnumber: u32, value: String, stops: &HashMap<String, u32>) {
    bars[barnumber as usize].stop = true;
    bars[barnumber as usize].stopvalue = stops[&value];
}
